### ISSUE TITLE
[2.3.1] Special Prices Showing at All Stores Level instead of Website Level

### ISSUE THREAD

**Issue by Sental** [magento/magento2#22567]
### Preconditions (*)
Magento 2.3.1
1. Set Price Scope to Website Level
2. Have 2 or More websites
3. Create a simple product that is attached to a store view on each website

### Steps to reproduce (*)
1. Set up the product price and special price on default/All store views level
2. add a different special price on a store view for a website. - See special price screenshots
3. View the frontend of the product on each store.

### Expected result (*)
1. You see the All store views special price on the unaltered store/website and the altered special price is available on that particular store.

### Actual result (*)
1. Both Websites show the all store views special price configuration - see frontend screenshots

Note
Complete re-indexing and flushing of the cache does not resolve the issue.
![special-price-all-stores](https://user-images.githubusercontent.com/35261502/57014584-b8f34680-6c08-11e9-80e7-7e0e5e0e17e1.PNG)
![special-price-website-1-uses-default](https://user-images.githubusercontent.com/35261502/57014585-b8f34680-6c08-11e9-9eea-1a6bf43cc8cd.PNG)
![special-price-website-2](https://user-images.githubusercontent.com/35261502/57014586-b8f34680-6c08-11e9-9b86-a62eba4dc06b.PNG)
![frontend-website-1-default-special-price](https://user-images.githubusercontent.com/35261502/57014592-c0b2eb00-6c08-11e9-93b7-da7169379b01.PNG)
![frontend-website-2-default-special-price](https://user-images.githubusercontent.com/35261502/57014593-c0b2eb00-6c08-11e9-952b-8d66c4794a03.PNG)

**Comment by rlavooij** [comment]
I also ran into this very same problem yesterday during a site upgrade.  
Reproduced on a clean magento 2.3.1 install. 

I also noticed, if you remove the special price from the global scope, and override it in the store views, it won't display anything "special" on the front-end of the site.

pricing entered looks to be fine in the catalog_product_index_price table too 
https://i.imgur.com/LFQEGLH.png

**Comment by rlavooij** [comment]
For a bit more visual explanation 2 screenshots with 2 cases.
left side is store view one
right side the other.

https://i.imgur.com/49oFlLb.png
in this case i have not set a default value for `Special Price`, only specified this in the store views


in the following screenshot i have set up the "all store views" value https://i.imgur.com/cjlLx96.png
then going back into the store views we can see that the special price is being pulled in from the all store views setting: https://i.imgur.com/IL2GHck.png

**Comment by ionutturdean** [comment]
Hi,
I have the same issue.
vendor/magento/module-catalog/Model/Product/Type/FrontSpecialPrice.php:117
I have noticed store zero is harcoded:
Original function:
private function isSuitableSpecialPrice($product, array $price): bool
    {
        $priceStoreId = $price[Store::STORE_ID];
        if (($priceStoreId == Store::DEFAULT_STORE_ID || $product->getStoreId() == $priceStoreId)
            && $price[SpecialPriceInterface::PRICE_FROM] == $product->getSpecialFromDate()
            && $price[SpecialPriceInterface::PRICE_TO] == $product->getSpecialToDate()) {
            return true;
        }
If I remove DEFAULT_STORE_ID is working:
Modified function:
private function isSuitableSpecialPrice($product, array $price): bool
    {
        $priceStoreId = $price[Store::STORE_ID];
        if (($product->getStoreId() == $priceStoreId)
            && $price[SpecialPriceInterface::PRICE_FROM] == $product->getSpecialFromDate()
            && $price[SpecialPriceInterface::PRICE_TO] == $product->getSpecialToDate()) {
            return true;
        }
Not sure is the right solution.

**Comment by Sental** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @Sental. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @Sental, here is your [Magento instance](https://i-22567-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-22567-2-3-develop.instances.magento-community.engineering/admin](https://i-22567-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by Sental** [comment]
@magento-engcom-team yes

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @shikhamis11 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-17191` were created

**Issue Available**: @shikhamis11, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by OvalMedia** [comment]
I am having the same issue on 2.2.9. [ionutturdean's fix](https://github.com/magento/magento2/issues/22567#issuecomment-494437836) works for me.

**Comment by mbernabeu** [comment]
same issue in 2.2.9

**Comment by scottsb** [comment]
It seems this was fixed in 2.3.2, but there is no fix applied to the 2.2.x branch, and since that branch is now closed to PRs, I don't think it's possible to fix. See commit 423f92c for the 2.3 fix.

**Comment by magento-engcom-team** [comment]
Hi @Sental. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-17191` was closed as `non-reproducible` in `2.3-develop`.
It means that Magento team either unable to reproduce this issue using provided _Steps to Reproduce_ from the **_Description section_** on clean Magento instance or the issue has been already fixed in the scope of other tasks.

But if you still run into this problem please update or provide additional information/steps/preconditions in the **_Description section_** and reopen this issue.

**Comment by LiamKarlMitchell** [comment]
Not sure if related but this issue came up in my google searches.
I had to go into the catalog_eav_attribute table and set the global to 2 for price, special price, special from date and special to date attributes.
Just setting the price to be website scoped in core_config table did not update these accordingly.


In my scenario I was adding a second store on a different url for consumer facing to a Magento 2 site that already had a business facing store.

```sql
-- Update special attributes to be website scoped.
UPDATE catalog_eav_attribute SET is_global = 2 WHERE attribute_id IN (SELECT attribute_id FROM eav_attribute WHERE attribute_code IN ('price','special_price', 'special_from_date', 'special_to_date'));
-- There is also a core_config setting for this found in Magento 2 admin under Catalog, Price which can be set to Global or Website scope.

-- Update existing historical special price to be store_id 1.
UPDATE catalog_product_entity_decimal SET store_id = 1 WHERE attribute_id = 78 AND store_id = 0;
-- Update existing historical special price dates to be on store_id 1.
UPDATE catalog_product_entity_datetime SET store_id = 1 WHERE attribute_id IN (79, 80) AND store_id = 0;
```

Just writing a comment in-case it helps someone else in the future.
